### PR TITLE
Refactor actions code and file structure

### PR DESCRIPTION
**Pull request by petar-qb**
## Description
The structure of the action files has been refactored according to several guidelines:
 - Remove as much non-VizroBaseModel stuff from `vizro/models/_action' as possible.
 - Gather all code related to the action loop mechanism and separate from Action model or predefined action functions.
 - Keep the Action model methods: `_get_components ` and `_make_action_callback ` close to the action loop engine mechanism implementation.
 - Keep `_actions_utils.py` close to predefined action functions.
 - Try to get rid of dynamic imports as much as possible

---

The file structure related to actions now looks like this:
```
📂 vizro/
    📂 actions/
        📄 __init__.py    ---    (only predefined actions _filter and _filter_interaction are exposed as public)
        📂 _callback_mappings/
            📄 __init__.py
            📄 _callback_mappings_utils.py    ---    (`_get_action_callback_mapping` utils)
            📄 _get_action_callback_mapping.py    ---    (only one function with pure dictionary mapping)
        📂 _action_loop/
            📄 __init__.py
            📄 _action_loop.py    ---    (ActionLoop class with _create_app_callbacks method that creates all action loop and action model callbacks and returns all the required components )
            📄 _get_action_loop_components.py    ---    (ex `_get_actions_components` function from `_callback_utils.py`)
            📄 _build_action_loop_callbacks.py    ---    (ex `_make_app_callback` function from `_callback_utils.py`)
        📂 predefined_functions/
            📄 __init__.py
            📄 _filter_action.py
            📄 _on_page_load_action.py
            📄 _parameter_action.py
            📄 _predefined_actions_utils.py    ---    (ex `_actions_utils.py`)
            📄 export_data_action.py
            📄 filter_interaction_action.py
    📂 models/
        📂 _action/
            📄 __init__.py    ---    (only the Action model is exposed as public)
            📄 _action.py    ---    (VizroBaseModel with build method: ex `_get_components` +  ex `_make_action_callback`)
            📄 _actions_chain.py    ---    (VizroBaseModel with `_action_validator_factory` - maybe we need to extract this validator as well)
```

---

Procedure for adding a new action:
    1. Add action function implementation as a separate file inside the `vizro/actions/` directory (like export_data.py).
    2. Update the `action_functions` mapping inside the `vizro/actions/__init__.py`.
    3. Update also the `__all__` in the same file if the action is public.
    4. Update the `action_callback_mapping` mapping inside the `vizro/actions/_callback_mappings/_get_action_callback_mapping.py`.

---

## Checklist

- [x] I have not referenced individuals, products or companies in any commits, directly or indirectly
- [x] I have not added data or restricted code in any commits, directly or indirectly
- [x] I have updated the docstring of any public function/class/model changed
- [ ] I have added the PR number to the change description in the changelog fragment, e.g. `Enable feature XXX (#123)` (if applicable)

## Types of changes

- [x] Docs/refactoring (non-breaking change which improves codebase)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: vizro-core/src/vizro/actions/_action_loop/_action_loop.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""The action manager handles creation of all required actions callbacks."""
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_action_loop.py:1]:
```suggestion
"""The action loop creates all the required action callbacks and its components."""
```

File: vizro-core/src/vizro/actions/_action_loop/_action_loop.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        Returns:
+            List of all required components for the action loop. e.g. List of dcc.Store, html.Div components.
+        """
+        _build_action_loop_callbacks()
+        return _get_action_loop_components()
+
+    @staticmethod
+    def _build_actions_models():
+        """Builds a callback for Action model and returns required components for the Action model callback to work.
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_action_loop.py:35]:
```suggestion
        """Builds a callback for each `Action` model and returns required components for these callbacks.
```

File: vizro-core/src/vizro/actions/_action_loop/_action_loop.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from vizro.models import Action
+
+
+class ActionLoop:
+    @classmethod
+    def _create_app_callbacks(cls) -> List[Union[dcc.Store, html.Div, dcc.Download]]:
+        """Builds callbacks for the action loop and for each Action in the Dashboard and returns their components.
+
+        Returns:
+            All required components for the action loop and for each Action in the Dashboard.
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_action_loop.py:19]:
```suggestion
            List of required components for the action loop and for each `Action` in the `Dashboard`.
```

File: vizro-core/src/vizro/actions/_action_loop/_action_loop.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        """
+        return cls._build_action_loop() + cls._build_actions_models()
+
+    @staticmethod
+    def _build_action_loop():
+        """Builds callbacks for the action loop and returns required components for the action loop mechanism to work.
+
+        Returns:
+            List of all required components for the action loop. e.g. List of dcc.Store, html.Div components.
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_action_loop.py:28]:
```suggestion
            List of required components for the action loop e.g. List[dcc.Store, html.Div].
```

File: vizro-core/src/vizro/actions/_action_loop/_action_loop.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        """
+        _build_action_loop_callbacks()
+        return _get_action_loop_components()
+
+    @staticmethod
+    def _build_actions_models():
+        """Builds a callback for Action model and returns required components for the Action model callback to work.
+
+        Returns:
+            List of all required components for each action model in the Dashboard. e.g. List of dcc.Download.
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_action_loop.py:38]:
```suggestion
            List of required components for each `Action` in the `Dashboard` e.g. List[dcc.Download]
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+from vizro._constants import ON_PAGE_LOAD_ACTION_PREFIX
+from vizro.actions import action_functions
+from vizro.managers import model_manager
+from vizro.models import Action
+from vizro.models._action._actions_chain import ActionsChain
+
+logger = logging.getLogger(__name__)
+
+
+# make callbacks that enable action loop mechanism to work
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:18]: Comment is redundant as you have it in the doc string already
```suggestion
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+def _build_action_loop_callbacks() -> None:
+    """Creates all required dash.callback for action loop."""
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:20]:
```suggestion
    """Creates all required dash callbacks for the action loop."""
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    # TODO - Reduce the number of the callbacks in the action loop mechanism
+    actions_chains = [actions_chain for _, actions_chain in model_manager._items_with_type(ActionsChain)]
+    actions = [action for _, action in model_manager._items_with_type(Action)]
+
+    gateway_triggers: List[Input] = []
+    for actions_chain in actions_chains:
+        # TODO: Seems like a good candidate for clientside callback
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:27]:
```suggestion
        # TODO: Potentially convert to clientside callback
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+    # TODO: don't create any components or callback if there's no actions configured
+    if not gateway_triggers:
+        gateway_triggers.append(Input("empty_input_store", "data"))
+
+    action_triggers = [Output({"type": "action_trigger", "action_name": action.id}, "data") for action in actions]
+    if not action_triggers:
+        action_triggers.append(Output("empty_output_store", "data", allow_duplicate=True))
+
+    # gateway
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:55]:
```suggestion
```

File: vizro-core/src/vizro/models/_action/__init__.py
Comment (maxschulz-COL) [inline vizro-core/src/vizro/models/_action/__init__.py:1]: is this file needed, do we not import the action model at the higher init?

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        prevent_initial_call=True,
+    )
+    def gateway(*gateway_triggers: List[dcc.Store]) -> List[Optional[str]]:
+        """Determines the final sequence of actions to be triggered.
+
+        Args:
+            gateway_triggers: Each 'gateway_trigger' (ctx.triggered_id) provides the 'id' (or trigger_id) from the
+                'actions_chain' that should be executed.
+
+
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:68]:
```suggestion
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+        Returns:
+            List of final action sequence names which need to be triggered in order.
+
+
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:72]:
```suggestion
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        Raises:
+            PreventUpdate:
+                If screen with triggers is rendered but component isn't triggered.
+        """
+        # Fetch all triggered action chain ids
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:77]: That's already clear from the variable name
```suggestion
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        logger.debug("=========== ACTION ===============")
+        logger.debug(f"Triggered component: {triggered_actions_chains_ids[0]}.")
+        final_action_sequence = [
+            {"Action ID": action.id, "Action name": action_functions[action.function._function]}
+            for action in model_manager[actions_chain_to_trigger].actions  # type: ignore[attr-defined]
+        ]
+        logger.debug(f"Actions to be executed as part of the triggered ActionsChain: {final_action_sequence}")
+        return [action_dict["Action ID"] for action_dict in final_action_sequence]
+
+    # update remaining actions
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:101]:
```suggestion
```

File: vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        # propagate sequence of actions from gateway callback
+        triggered_id = ctx.triggered_id
+        if triggered_id == "set_remaining":
+            return set_remaining
+        # pop first action
+        if triggered_id == "cycle_breaker_div":
+            return remaining_actions[1:]
+        return []
+
+    # executor
```
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/actions/_action_loop/_build_action_loop_callbacks.py:135]:
```suggestion
```

**Review by maxschulz-COL — COMMENTED** [review]
This looks good from my side. I did not go into the details anymore, but given our discussions and the review of the previous PR I think we should be good to go. Have one remaining question about the `init` file though

**Review by huong-li-nguyen — APPROVED** [review]
Looks good from my side! It looks a lot cleaner, have some minor comments on rephrasing.

Also can you double-check your comments? I've suggested to remove some if the variable/function names or the docstrings made it clear already. Don't know if I caught all of them, but feel free to remove comments if they don't much information

File: vizro-core/src/vizro/models/_action/__init__.py
Comment (petar-qb) [reply vizro-core/src/vizro/models/_action/__init__.py:1]: Content of the higher init (vizro-core/src/vizro/models/__init__.py) is accordingly changed
from
`from ._action._action import Action`
to
`from ._action import Action`

I prefer this approach because the __init__ files don't go deep into the file structure. But I see that there are several different ways of how we implemented the import system in different init files. So we probably need to re-consolidate it in different PR.

Comment (maxschulz-COL) [reply vizro-core/src/vizro/models/_action/__init__.py:1]: OK I see, fine for me either way.

**Review by huong-li-nguyen — APPROVED** [review]

**Review by maxschulz-COL — APPROVED** [review]
🚀
